examples/ivr_outbound_dialer.py

```diff
@@ -0,0 +1,409 @@
+        # Developer would start the introductory IVR message
+        if sess.is_outbound():
+
+            # Start recording a call
+            call.vars['record'] = True
+            sess.start_record(
+                '{}/callee_{}.wav'.format(self.recdir, sess.uuid),
+                stereo=self.stereo
+            )
+
+            # Play IVR initial greeting
+            call.vars['play_welcome'] = True
+            play_filename = '{}/ivr-welcome.wav'.format(self.sound_dir)
+            sess.playback(play_filename)
+
+            # create a list for storing received dtmf digits
+            call.vars['incoming_dtmf'] = []
+
+            # At this point we wait for the PLAYBACK_STOP event to
+            # arrive and then start a timeout trigger as in DTMF handler
+
+    @staticmethod
+    def cancel_timer(sess):
+        timer = sess.vars.get('dtmf_timer')
+        if timer:
+            log.debug("'{}': Cancel dtmf timeout job".format(sess.uuid))
+            timer.cancel()  # if timer in 'initial' state this is a noop
+        return timer
+
+    def start_dtmf_timer(self, sess, timeout=3):
+        """Start a timer thread which will invoke the dtmf timeout handler after
+        `timeout` seconds. The default interval is 3 seconds.
+        """
+        timer = self.cancel_timer(sess)
+
+        # allocate a new timer
+        timer = threading.Timer(timeout, self.dtmf_timeout_action, [sess])
+        timer.daemon = True  # thread dies with parent process
+        timer.start()  # NOTE: this launches a thread per call
+        sess.vars['dtmf_timer'] = timer
+        return timer
+
+    def dtmf_timeout_action(self, sess):
+        """Timer handler that implements DTMF timeout
+        """
+        call = sess.call
+        log.info("'{}': DTMF timeout".format(sess.uuid))
+
+        if call.vars.get('playing') is True:
+            call.vars['playing'] = False
+            sess.breakmedia()  # stop playback
+
+        # Reset incoming dtmf queue
+        call.vars['incoming_dtmf'] = []
+
+        # Example of playing a prompt urging the user to try again
+        play_filename = '{}/ivr-hello.wav'.format(self.sound_dir)
+        call.vars['playing'] = True
+        sess.playback(play_filename)
+
+        # Trigger dtmf timeout again
+        self.start_dtmf_timer(sess)
+
+    @event_callback('DTMF')
+    def on_digit(self, sess):
+        """Process DTMF digit events
+        """
+        call = sess.call
+        digit = sess['DTMF-Digit']
+        log.info("'{}': DTMF dtmf digit '{}'".format(sess.uuid, digit))
+
+        # Add incoming digit into the digit queue
+        call.vars.setdefault('incoming_dtmf', []).append(digit)
+
+        # DTMF has just been detected, stop playing any files to the user
+        if call.vars.get('playing') is True:
+            sess.breakmedia()
+
+        # Stop the dtmf timeout timer if one exists
+        self.cancel_timer(sess)
+
+        # IVR Menu - a map of digits to actions
+        actions = {
```

Comment:
I was thinking this map's values might be better suites as functions which are defined by the user ahead of time?